src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,270 @@
+
+    public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+        => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+    private static readonly DiagnosticDescriptor s_descriptor = CreateDescriptorWithId(
+        IDEDiagnosticIds.UseCollectionExpressionForArrayDiagnosticId,
+        EnforceOnBuildValues.UseCollectionExpressionForArray,
+        new LocalizableResourceString(nameof(AnalyzersResources.Simplify_collection_initialization), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)),
+        new LocalizableResourceString(nameof(AnalyzersResources.Collection_initialization_can_be_simplified), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)),
+        isUnnecessary: false);
+
+    private static readonly DiagnosticDescriptor s_unnecessaryCodeDescriptor = CreateDescriptorWithId(
+        IDEDiagnosticIds.UseCollectionExpressionForArrayDiagnosticId,
+        EnforceOnBuildValues.UseCollectionExpressionForArray,
+        new LocalizableResourceString(nameof(AnalyzersResources.Simplify_collection_initialization), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)),
+        new LocalizableResourceString(nameof(AnalyzersResources.Collection_initialization_can_be_simplified), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)),
+        isUnnecessary: true);
+
+    public CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer()
+        : base(ImmutableDictionary<DiagnosticDescriptor, IOption2>.Empty
+                .Add(s_descriptor, CodeStyleOptions2.PreferCollectionExpressionForArray)
+                .Add(s_unnecessaryCodeDescriptor, CodeStyleOptions2.PreferCollectionExpressionForArray))
+    {
+    }
+
+    protected override void InitializeWorker(AnalysisContext context)
+        => context.RegisterCompilationStartAction(OnCompilationStart);
+
+    private void OnCompilationStart(CompilationStartAnalysisContext context)
+    {
+        if (!context.Compilation.LanguageVersion().IsCSharp12OrAbove())
+            return;
+
+        // We wrap the SyntaxNodeAction within a CodeBlockStartAction, which allows us to
+        // get callbacks for object creation expression nodes, but analyze nodes across the entire code block
+        // and eventually report fading diagnostics with location outside this node.
+        // Without the containing CodeBlockStartAction, our reported diagnostic would be classified
+        // as a non-local diagnostic and would not participate in lightbulb for computing code fixes.
+        context.RegisterCodeBlockStartAction<SyntaxKind>(context =>
+        {
+            context.RegisterSyntaxNodeAction(
+                context => AnalyzeArrayInitializer(context),
+                SyntaxKind.ArrayInitializerExpression);
+        });
+    }
+
+    private static bool IsInTargetTypedLocation(SemanticModel semanticModel, ExpressionSyntax expression, CancellationToken cancellationToken)
```

Comment:
Will this suffer from many semantic model calls? I think the syntactic filteration would still have many applicable nodes, and hence many semantic model calls.

Reply:
Not sure what thsi question means.

Reply:
I mean, does this make too many semantic model caĺls that could affect performance?

Reply:
I don't see anything that would cause too many semantic model calls here.  We effectively are doing a check right on the item of interest.